[Misra Aditya] iP

MadLamprey opened the pull request:
# Bruno: Your Productivity Companion 🐶

> "If I could be half the person my dog is, I'd be twice the human I am." - Charles Yu [(source)](https://www.goodreads.com/quotes/488888-if-i-could-be-half-the-person-my-dog-is)

Bruno helps you remember to complete all your tasks, in order to maximise your productivity. 

To make Bruno yours,
- Download the ```.jar``` file [(bruno.jar)](https://github.com/MadLamprey/ip/releases/tag/v0.1)
- Run ```java -jar bruno.jar``` in the directory where the file is present
- Make full use of Bruno to maximise your productivity!

Features:
1. Add tasks based on type - ```ToDo```, ```Deadline``` or ```Event```
2. Save and load task list from a ```.txt``` file
3. Mark/unmark tasks on completion
4. View schedule for a particular date
5. Find tasks based on a keyword

What makes Bruno the best?

- [x] text-based
- [x] easy-to-learn with simple commands
- [x] ~FAST~ SUPER FAST to use

And, it is **FREE** to use!

Here's the code that gets Bruno barking!

```java
public static void main(String[] args) {
        Bruno bruno = new Bruno("data/", "bruno.txt");
        bruno.runBruno();
}

```

j-lum committed:
Merge pull request #3 from j-lum/inline-javafx-tutorials

Inline JavaFX tutorials

j-lum committed:
Update Gradle tutorial to reflect new checkstyle config

j-lum committed:
Fix typo referring to `HelloWorld`

j-lum committed:
Add images to JavaFX tutorial 3

j-lum committed:
Merge pull request #7 from j-lum/tutorial-fix

Address issues #5 and #6

j-lum committed:
Remove references to DukeStub

j-lum committed:
Merge pull request #8 from j-lum/tutorial-fix

Remove references to DukeStub

j-lum committed:
Change indentation in FXML samples to match AB-3

j-lum committed:
Merge pull request #14 from j-lum/fxml-indent

Change indentation in FXML samples to match AB-3

j-lum committed:
JavaFX tutorial: Support cross-platform JARs

The OpenJFX plugin expects applications to be modular and bundled
with jlink, resulting in fat jars that are not cross-platform. Let's
manually include the required dependencies so that shadow can package
them properly.

j-lum committed:
Merge pull request #35 from j-lum/xplatformjfx

JavaFX tutorial: Support cross-platform JARs

jiachen247 committed:
Remove unnecessary override annotation

Remove extra '@Override' at the start of the code example.

j-lum committed:
Merge pull request #41 from jiachen247/patch-2

Remove unnecessary override annotation

Commit:
Remove generic tutorials

Let's remove generic tutorials from this repo as they are not specific
to this project tempalte

Commit:
Add text-ui-test files

Commit:
Add Gradle support

damithc committed:
README.md: Fix typo of => or

Commit:
Improve runtest.* scripts

* Improve the javac command to work for multiple java files.
* In runtest.bat, delete ACTUAL.TXT only if it exists.

Commit:
README: Update setup instructions to match the latest IDE version

Commit:
README.md: Tweak template text

* Let's replace 'Feature 1' with 'Feature-ABC'
  reason: some students kept the 'Feature 1' in the final UG.
* Let's tweak the template for features and usage
  reasons: to fix grammar, to distinguish between output and outcome

damithc committed:
README: Mention the project language level setting

seanleong339 committed:
.gitignore: Fix ACTUAL.txt -> ACTUAL.TXT  (#73)

text-ui-test/runtest.bat|sh scripts generate a file ACTUAL.TXT.
However, .gitignore uses ACTUAL.txt, which means the generated
file will not be ignored by Git on non-Windows OS.

Let's update .gitignore as ACTUAL.txt -> ACTUAL.TXT

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Initial Version

Commit:
Level 1 Updates

Commit:
Level 2 Updates

Commit:
Level 3 Updates [Marking and Unmarking]

Commit:
Level 4 Updates [Different Task Types]

Commit:
Set Up Automated UI Testing

Commit:
Error Handling

Commit:
Exception Handling

Commit:
Delete Feature

Commit:
Collections

Commit:
Enums

Commit:
Add functionality to save tasks locally

Commit:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add functionality to save tasks locally

Commit:
Add Date and Time for Deadlines and Events

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add Date and Time for Deadlines and Events

Commit:
Add more OOP

Commit:
Merge commit 'e6cbdccc35c090d93883d9ff27f3e809a02b4b1a'

* commit 'e6cbdccc35c090d93883d9ff27f3e809a02b4b1a':
  Add more OOP

Commit:
Fix bugs and change style

Commit:
Refactor classes into packages

Commit:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Refactor classes into packages

Commit:
Merge branch 'add-gradle-support' of https://github.com/MadLamprey/ip

Commit:
Add Gradle support and fix style issues

Commit:
Merge branch 'branch-A-Gradle'

* branch-A-Gradle:
  Add Gradle support and fix style issues

Commit:
Add JUnit tests

Commit:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  Add JUnit tests

Commit:
Create Jar file and fix bugs

Commit:
Merge branch 'branch-A-Jar'

* branch-A-Jar:
  Create Jar file and fix bugs

Commit:
Fix bugs

Commit:
Rename UITest file to match conventions

Commit:
Add JavaDoc comments to methods and classes

Commit:
Fix code to match coding standard

Commit:
Fix issue with saving to file

Commit:
Add functionality to find tasks

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments to methods and classes

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fix code to match coding standard

# Conflicts:
#	src/main/java/bruno/Bruno.java
#	src/main/java/bruno/Parser.java
#	src/main/java/bruno/Storage.java
#	src/main/java/bruno/TaskList.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add functionality to find tasks

# Conflicts:
#	src/main/java/bruno/Storage.java
#	src/main/java/bruno/TaskList.java
#	src/test/java/bruno/StorageTest.java
#	src/test/java/bruno/TaskListTest.java

Commit:
Fix CheckStyle violations

Commit:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Fix CheckStyle violations

winson8222 commented on src/main/java/bruno/TaskList.java:
```diff
@@ -0,0 +1,260 @@
+     */
+    public void addToDo(String task) throws BrunoException {
+        if (task.split(" ").length == 1) {
+            System.out.print("\t");
```
Perhaps it will be good to make a function to display messages together with \t so that `System.out.print("\t")` do not have to be repeated used.

winson8222 commented on src/main/java/bruno/UI.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Displays the divider lines before and after each command.
+     */
+    public void displayLines() {
```
Maybe the line string can be a constant?

winson8222 commented on src/test/java/bruno/ParserTest.java:
```diff
@@ -0,0 +1,43 @@
+        parser = new Parser(new TaskList(new Storage("xxx", "yyy"), new UI()));
+    }
+
+    @Test void testParse_normalInput_correctOutputGenerated() {
```
The naming of the test methods followed the coding standard guide!

winson8222 commented on src/main/java/bruno/Parser.java:
```diff
@@ -0,0 +1,70 @@
+    public boolean parseInput(String input) throws BrunoException {
+        try {
+            TaskType type = TaskType.valueOf(input.split(" ")[0].toUpperCase());
+            switch (type) {
```
Switch statements have no indentation for cases. Good job!

winson8222 commented on src/test/java/bruno/task/DeadlineTest.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Test
+    void testGetString() {
```
Perhaps can specify what is the input string, since this is a special case of date string

winson8222 reviewed (commented):
This is very well structure and documented. Coding standard is largele adhered to.

Commit:
Add GUI for application

Commit:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI for application

Commit:
Fix bugs

evanyan13 commented on src/main/java/bruno/TaskList.java:
```diff
@@ -0,0 +1,255 @@
+        if (!task.substring(task.indexOf("deadline") + 1).contains("/by")) {
+            throw new BrunoMissingDeadlineException();
+        }
+        tasks.add(new Deadline(task.substring(task.indexOf(' ') + 1, task.indexOf('/') - 1),
```
for all the `addTask()` methods in this class, maybe can consider extracting out the parsing methods like `substring()` and `indexOf()` to make the code more readable?

```suggestion
        String description = task.substring(task.indexOf(' ') + 1, task.indexOf('/') - 1);
        String by = task.substring(task.lastIndexOf('/') + 4);
        tasks.add(new Deadline(description, by));
```

evanyan13 commented on src/main/java/bruno/TaskList.java:
```diff
@@ -0,0 +1,255 @@
+            throw new BrunoIndexOutOfBoundsException("mark");
+        }
+        if (Integer.parseInt(markVal) < 0) {
+            throw new BrunoNegativeArgException("mark");
```
I like how you have 2 different exceptions to be thrown for either out of bound or negative index 👍

evanyan13 commented on src/main/java/bruno/TaskList.java:
```diff
@@ -0,0 +1,255 @@
+    public String displayListSum() {
+        String taskInfo = "Now you have " + tasks.size() + (tasks.size() == 1 ? " task" : " tasks") + " "
+                + "in your "
+                + "tasks.";
```
Do we need the indentation to be like this? Maybe the indentation came when you accidentally edited this line?
```suggestion
                String taskInfo = "Now you have " + tasks.size() 
                            + (tasks.size() == 1 ? " task" : " tasks") + " in your tasks.";
```

evanyan13 commented on src/main/java/bruno/UI.java:
```diff
@@ -0,0 +1,42 @@
+     */
+    public String displayGreeting() {
+        String name = "Bruno";
+        return "Woof Woof! I'm " + name + " 🐾" + "\nHow can I help you?";
```
Cute emotion! I'm not sure if this is a standard but I have this habit of putting "\n" at the end of the line instead of at the start of next line. Even if both are same effect I guess 😆 
```suggestion
        return "Woof Woof! I'm " + name + " 🐾 \n" + "How can I help you?";
```

evanyan13 commented on src/main/java/bruno/task/Task.java:
```diff
@@ -0,0 +1,89 @@
+     * @throws DateTimeException Thrown if string cannot be converted to LocalDateTime.
+     */
+    public LocalDateTime convertToLocalDateTime(String s) throws DateTimeException {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Just a quick idea, what if the user do not want to enter this DateTime format or wish to enter another format? Maybe can think about how to include a few more common DateTime format.

evanyan13 reviewed (commented):
LGTM 👍  Just some nits I have raised in my comments. Some might not be coding standard but just suggestions that you can consider to implement?

Commit:
Add note functionality to tasks

Sometimes additional details have to be added to the tasks.

A new command, 'NOTE' is added which allows addition of notes to tasks.

Notes can be added either at the time of creation of a new task, or at a later stage, using the 'note' command.

Commit:
Merge branch 'branch-BCD-extension'

* branch-BCD-extension:
  Add note functionality to tasks

Commit:
Add assertions to verify code

Commit:
Refactor code to improve code quality

MadLamprey committed:
Merge pull request #1 from MadLamprey/branch-A-CodeQuality

Refactor code to improve code quality

Commit:
Merge branch 'master' of https://github.com/MadLamprey/ip

* 'master' of https://github.com/MadLamprey/ip:
  Refactor code to improve code quality

Commit:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assertions to verify code

# Conflicts:
#	src/main/java/bruno/Main.java
#	src/main/java/bruno/Storage.java
#	src/main/java/bruno/TaskList.java

Commit:
Integrate CI

MadLamprey committed:
Create gradle.yml

MadLamprey committed:
Update gradle.yml

Commit:
Fix bugs

Commit:
Fix bugs raised in CI

Commit:
Add User Guide

Commit:
Merge branch 'branch-A-UserGuide'

* branch-A-UserGuide:
  Add User Guide

Commit:
Add more JUnit Tests

Commit:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  Add more JUnit Tests

MadLamprey committed:
Update README.md

Commit:
Merge branch 'master' of https://github.com/MadLamprey/ip

* 'master' of https://github.com/MadLamprey/ip:
  Update README.md

Commit:
Update build.gradle

Commit:
Update README.md

Commit:
Update build.gradle

MadLamprey committed:
Create _config.yml

MadLamprey committed:
Update  _config.yaml

Commit:
Merge branch 'master' of https://github.com/MadLamprey/ip

* 'master' of https://github.com/MadLamprey/ip:
  Update  _config.yaml
  Create _config.yml

Commit:
Fix bugs

MadLamprey committed:
Add agent mode

MadLamprey committed:
Add AI-powered increment reviewer workflow and script

- Add GitHub Actions workflow for automated code reviews on tag pushes
- Add Python script that fetches requirements, generates diffs, and posts LLM reviews
- Supports all Level-*, A-*, and B-* increment tags
- Uses Groq API for AI-powered code review generation

MadLamprey committed:
Fix A-JUnit anchor ID to match actual webpage structure

- Updated duke-a-junit-add-junit-tests to duke-a-junit
- Requirements fetching now works correctly

MadLamprey committed:
Fix all incorrect anchor IDs in TAG_REQUIREMENTS mapping

- Updated all A-* increment anchors to match actual webpage structure
- A-MoreOOP: duke-a-moreoop-use-more-oop → duke-a-moreoop
- A-Packages: duke-a-packages-organize-into-packages → duke-a-packages
- A-Gradle: duke-a-gradle-use-gradle → duke-a-gradle
- A-JUnit: duke-a-junit-add-junit-tests → duke-a-junit
- A-Jar: duke-a-jar-create-a-jar-file → duke-a-jar
- A-JavaDoc: duke-a-javadoc-javadoc → duke-a-javadoc
- A-CodingStandard: duke-a-codingstandard-follow-the-coding-standard → duke-a-codingstandard

MadLamprey committed:
Fix A-Gradle anchor ID to match actual webpage structure

- Corrected duke-a-gradle to duke-a-gradle-use-gradle

MadLamprey committed:
Update documentation

MadLamprey committed:
Update knowledge base to be static

MadLamprey committed:
Remove legacy code

MadLamprey committed:
Update model

MadLamprey committed:
Debugging code

MadLamprey committed:
Fix bugs

MadLamprey committed:
Add debugging code

MadLamprey committed:
Add debugging code

MadLamprey committed:
Fix bugs

MadLamprey committed:
Refactor increment reviewer to issue-based flow

MadLamprey committed:
Add small change for reviewer test

MadLamprey committed:
Add small change for reviewer test

MadLamprey committed:
Switch to v2 reviewer with three modes (adversarial, rubric, quality)

cs72-dev committed:
Reviewer: empty env-vars use defaults; constrain adversarial JSON

cs72-dev committed:
Reviewer: smaller prompt budgets and graceful 413 handling for /rubric and /quality
